kahuna/public/js/services/label.js

```diff
@@ -2,18 +2,43 @@ import angular from 'angular';
 
 var labelService = angular.module('kahuna.services.label', []);
 
-labelService.factory('labelService', ['$rootScope', '$q', function ($rootScope, $q) {
+labelService.factory('labelService',
+                     ['$rootScope', '$q', 'apiPoll', 'imageAccessor',
+                      function ($rootScope, $q, apiPoll, imageAccessor) {
+
+    function readLabelName(label) {
+        return label.data;
+    }
+
+    function readLabelsName(labels) {
+        return labels.map(readLabelName);
+    }
+
+    function labelsEquals(labelsA, labelsB) {
+        return angular.equals(readLabelsName(labelsA), readLabelsName(labelsB));
+    }
+
+    function untilLabelsEqual(image, expectedLabels) {
+        return image.get().then(apiImage => {
+            const apiLabels = imageAccessor.readLabels(apiImage);
+            if (labelsEquals(apiLabels, expectedLabels)) {
+                return apiImage;
+            } else {
+                return $q.reject();
+            }
+        });
+    }
+
     function remove (image, label) {
-        var existingLabels = image.data.userMetadata.data.labels.data;
+        var existingLabels = imageAccessor.readLabels(image);
         var labelIndex = existingLabels.findIndex(lbl => lbl.data === label);
-
         if (labelIndex !== -1) {
             return existingLabels[labelIndex]
                 .delete()
-                .then(newLabels => {
-                    image.data.userMetadata.data.labels = newLabels;
-                    $rootScope.$emit('image-updated', image, image);
-                    return image;
+                .then(newLabels => apiPoll(() => untilLabelsEqual(image, newLabels.data)))
```

Comment:
We need to avoid mutating image resources in place so the object identity can be used to compare two versions of an image resource (e.g. [image objects in the results collection](https://github.com/guardian/grid/pull/1200/files#diff-04b33f0ec2971bf016d5d80534267132R102)). Basically mixing an immutable model with mutable objects defeats all optimisations and capacity to reason about changes. We should move toward more immutability.

That said we should be able to do this without having to rely on the server to hand us fresh image resource objects. If modifying an image (or sub resources of an image), we should be able to update our internal state ("the state of that sub resource and its parent image resource") without polling the API. The server said the update was successful so we can reflect it onto our in-memory model without doing a read-after-write, which is an anti-pattern.

Happy to talk more about ideas around this.